plugins/inputs/nsdp/nsdp.go

```diff
@@ -0,0 +1,132 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package nsdp
+
+import (
+	_ "embed"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/tdrn-org/go-nsdp"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type NSDP struct {
+	Address     string          `toml:"address"`
+	DeviceLimit uint            `toml:"device_limit"`
+	Timeout     config.Duration `toml:"timeout"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	conn *nsdp.Conn
+}
+
+func (*NSDP) SampleConfig() string {
+	return sampleConfig
+}
+
+func (n *NSDP) Init() error {
+	if n.Address == "" {
+		n.Address = nsdp.IPv4BroadcastTarget
+	}
+	if n.Timeout <= 0 {
+		n.Timeout = config.Duration(2 * time.Second)
+	}
+	return nil
+}
+
+func (n *NSDP) Start(telegraf.Accumulator) error {
+	conn, err := nsdp.NewConn(n.Address, n.Log.Level().Includes(telegraf.Trace))
+	if err != nil {
+		return fmt.Errorf("failed to create connection to address %s: %s", n.Address, err)
+	}
+	conn.ReceiveDeviceLimit = n.DeviceLimit
+	conn.ReceiveTimeout = time.Duration(n.Timeout)
+	n.conn = conn
+	return nil
+}
+
+func (n *NSDP) Stop() {
+	if n.conn == nil {
+		return
+	}
+	n.conn.Close()
+	n.conn = nil
+}
+
+func (n *NSDP) Gather(acc telegraf.Accumulator) error {
+	var err error
+	if n.conn == nil {
+		err = n.Start(nil)
+	}
+	if err != nil {
+		return err
+	}
+	request := nsdp.NewMessage(nsdp.ReadRequest)
+	request.AppendTLV(nsdp.EmptyDeviceModel())
+	request.AppendTLV(nsdp.EmptyDeviceName())
+	request.AppendTLV(nsdp.EmptyDeviceIP())
+	request.AppendTLV(nsdp.EmptyPortStatistic())
+	responses, err := n.conn.SendReceiveMessage(request)
+	if err != nil {
+		// Close malfunctioning connection and re-connect on next Gather call
+		n.Stop()
+		return fmt.Errorf("failed to query address %s: %w", n.Address, err)
+	}
+	for device, response := range responses {
+		n.Log.Tracef("Processing device: %s", device)
+		n.gatherDevice(acc, device, response)
+	}
```

Comment:
How about adding some context with comments?
```suggestion

	// Send a request to discover all devices
	request := nsdp.NewMessage(nsdp.ReadRequest)
	request.AppendTLV(nsdp.EmptyDeviceModel())
	request.AppendTLV(nsdp.EmptyDeviceName())
	request.AppendTLV(nsdp.EmptyDeviceIP())
	request.AppendTLV(nsdp.EmptyPortStatistic())
	responses, err := n.conn.SendReceiveMessage(request)
	if err != nil {
		// Close malfunctioning connection and re-connect on next Gather call
		n.Stop()
		return fmt.Errorf("failed to query address %s: %w", n.Address, err)
	}
	
	// Convert the discovered devices to metrics
	for device, response := range responses {
		n.Log.Tracef("Processing device: %s", device)
		n.gatherDevice(acc, device, response)
	}
```

Reply:
Adapted via 284fc600122546d38c8062a4f38bae035eb6dc93